rust/signed_doc/src/lib.rs

```diff
@@ -105,6 +106,177 @@ impl CatalystSignedDocument {
+                error_report.functional_validation(
+                    &format!("Document Election ID {election_id:?} is invalid"),
+                    "During Document Election ID UUID verification",
+                );
+            }
+        }
+
+        if let Some(category_id) = extra.category_id() {
+            if !category_id.is_valid() {
+                error_report.functional_validation(
+                    &format!("Document Category ID {category_id:?} is invalid"),
+                    "During Document Category ID UUID verification",
+                );
+            }
+        }
+
+        match self.as_cose_sign() {
+            Ok(cose_sign) => {
+                let signatures = self.signatures().cose_signatures();
+                for (idx, kid) in self.signatures().kids().iter().enumerate() {
+                    let pk = pk_getter(kid);
+                    let signature = &signatures[idx];
+                    let tbs_data = cose_sign.tbs_data(&[], signature);
+                    match signature.signature.as_slice().try_into() {
+                        Ok(signature_bytes) => {
+                            let signature = ed25519_dalek::Signature::from_bytes(signature_bytes);
+                            if let Err(e) = pk.verify_strict(&tbs_data, &signature) {
+                                error_report.functional_validation(
+                                    &format!(
+                                        "Verification failed for signature with Key ID {kid}: {e}"
+                                    ),
+                                    "During signature validation with verifying key",
+                                );
+                            }
+                        },
+                        Err(_) => {
+                            error_report.invalid_value(
+                                "cose signature",
+                                &format!("{}", signature.signature.len()),
+                                &format!("must be {}", ed25519_dalek::Signature::BYTE_SIZE),
+                                "During encoding cose signature to bytes",
+                            );
+                        },
+                    }
+                }
+            },
+            Err(e) => {
+                error_report.other(
+                    &format!("{e}"),
+                    "During encoding signed document as COSE SIGN",
+                );
+            },
+        }
+
+        if error_report.is_problematic() {
+            return Err(CatalystSignedDocError::new(
+                error_report,
+                anyhow::anyhow!("Verification failed for Catalyst Signed Document"),
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// Returns a signed document `Builder` pre-loaded with the current signed document's
+    /// data.
+    #[must_use]
+    pub fn as_signed_doc_builder(&self) -> Builder {
+        Builder {
+            metadata: Some(self.inner.metadata.clone()),
```

Comment:
Seems it would be better to make all `Builder` fields private and just call `with_metadata`, `with_signatures` etc.